### DISCUSSION TITLE
Best practise to start a new repo

### DISCUSSION THREAD

**Discussion by FTamas77** [community/community#23617] [New to GitHub]
<p>Hi, I’ve a git repo. I would like to use this repo for two another project. Could you givve me some hint what the best practise is? I simply create a two new project and copy the source and loose the histroy or I use fork twice? But in the last case I cannot have a totally new repo with a new name? I could use branching, but it is not for totally different projects? I could create and publich a package, but I don’t want to create a package, because it is an additional work. Maybe a sub-branch is better than a unnecessary package.</p>
<p>I have a good basic low layer project and I would like to use it in other project. Thanks,</p>

**Comment by jfullstackdev** [comment, marked as answer]
<aside class="quote no-group" data-username="FTamas77" data-post="1" data-topic="219653">
<div class="title">
<div class="quote-controls"></div>
 FTamas77:</div>
<blockquote>
<p>Maybe a sub-branch is better than a unnecessary package.</p>
</blockquote>
</aside>
<p>do you mean they are two projects but you need to use the other repo inside the other repo?</p>
<p>if that is the case, it’s submodule, check this out,</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/2016-02-01-working-with-submodules/">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">

      <a href="https://github.blog/2016-02-01-working-with-submodules/" target="_blank" rel="noopener nofollow ugc" title="03:52AM - 02 February 2016">The GitHub Blog – 2 Feb 16</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/361;"><img src="https://user-images.githubusercontent.com/47092464/181078248-167844f6-04f9-477e-8d16-80084db7bbcd.png" class="thumbnail" width="690" height="361" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c017c230d6e8ab7ae3a76b88d5272b38b442726f_2_690x361.png, https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c017c230d6e8ab7ae3a76b88d5272b38b442726f_2_1035x541.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/c/0/c017c230d6e8ab7ae3a76b88d5272b38b442726f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/0/c017c230d6e8ab7ae3a76b88d5272b38b442726f_2_10x10.png"></div>

<h3><a href="https://github.blog/2016-02-01-working-with-submodules/" target="_blank" rel="noopener nofollow ugc">Working with submodules | The GitHub Blog</a></h3>

  <p>Eventually, any interesting software project will come to depend on another project, library, or framework. Git provides submodules to help with this. Submodules allow you to include or embed one or more repositories as a</p>

  <p>
    <span class="label1">Est. reading time: 7 minutes</span>
  </p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>for best practices, as for me, I only use one project per repo,</p>

**Comment by FTamas77** [comment]
<p>Thank you very much your reply.<br>
Maybe submodules a good solution in my case (I liked the link :slight_smile: ).<br>
What I want?<br>
Create a repo with OpenCL, Test packages and everything…<br>
Later I would like to create a repo or project for image processing and a different one for signal processing. I don’t want to copy or write again “hello world” OpenCL application/repo.<br>
Thanks to your link, I guess the sub-repo seems a good way.</p>

**Comment by FTamas77** [comment]
<p>I’ve added a submodule and I also push.<br>
Now it is under version handler. Is it OK?<br>
It is here:</p><aside class="onebox githubfolder" data-onebox-src="https://github.com/FTamas77/Demo/tree/develop/depedencies">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/FTamas77/Demo/tree/develop/depedencies" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h3><a href="https://github.com/FTamas77/Demo/tree/develop/depedencies" target="_blank" rel="noopener nofollow ugc">Demo/depedencies at develop · FTamas77/Demo</a></h3>

  <p><a href="https://github.com/FTamas77/Demo/tree/develop/depedencies" target="_blank" rel="noopener nofollow ugc">develop/depedencies</a></p>

  <p><span class="label1">Contribute to FTamas77/Demo development by creating an account on GitHub.</span></p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
How I can update (to be more precise, I want to handle versioning myself).<br>
Why is it under my version handler?<br>
I read about it on the internet and unfortunately it is still not clear.<br>
Anyway, I’m happy now, because I can integrate into my CMake system.<br>
But a little bit more help would be nice. Thanky my friend!</p>
<p>(By the way, your suggestion is the best as you can see.)</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="FTamas77" data-post="4" data-topic="219653">
<div class="title">
<div class="quote-controls"></div>
 FTamas77:</div>
<blockquote>
<p>How I can update (to be more precise, I want to handle versioning myself).<br>
Why is it under my version handler?</p>
</blockquote>
</aside>
<p>A submodule always points at a specific commit in the referenced repository. You can go into the submodule directory and pull updates (or check out any commit/ref) like you would in an independent repository. If you do, the surrounding repository (yours) will show that as a change you can commit when you’re ready.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="FTamas77" data-post="6" data-topic="219653">
<div class="title">
<div class="quote-controls"></div>
 FTamas77:</div>
<blockquote>
<p>2.) How I can update submodule?<br>
3.)I remember, we had to “include” submodule<br>
4.) What if I change something on that repo?</p>
</blockquote>
</aside>
<p>I answered this above: The directory with the submodule is a repository of its own, you can work with it like any other repository. If the HEAD commit in the submodule changes your surrounding repository will show a change.</p>
<aside class="quote no-group" data-username="FTamas77" data-post="6" data-topic="219653">
<div class="title">
<div class="quote-controls"></div>
 FTamas77:</div>
<blockquote>
<p>I know I can google for it. But your explanation is much better.</p>
</blockquote>
</aside>
<p>There are people here who are happy to answer questions (including me), but not even trying to figure things out yourself is disrespectful. Remember, most people here (with the exception of a few GitHub staff members) are just nice people trying to help in their free time. A good starting point for anything Git related is the <a href="https://git-scm.com/book/en/v2" class="inline-onebox">Git - Book</a>. If you get stuck or are uncertain, ask, and provide the necessary details.</p>
